url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/SOCServer.java

```diff
@@ -0,0 +1,9048 @@
+                        // send at most 1 message to each connected client, split by client version.
+
+                        final int cversMax = getMaxConnectedCliVersion();
+                        int newgameMaxCliVers;
+                        if ((gaOpts != null) && (cversMax >= SOCNewGameWithOptions.VERSION_FOR_NEWGAMEWITHOPTIONS))
+                        {
+                            broadcastToVers
+                                (SOCNewGameWithOptions.toCmd(gaName, gaOpts, gVers),
+                                 SOCNewGameWithOptions.VERSION_FOR_NEWGAMEWITHOPTIONS, Integer.MAX_VALUE);
+                            newgameMaxCliVers = SOCNewGameWithOptions.VERSION_FOR_NEWGAMEWITHOPTIONS - 1;
+                        } else {
+                            newgameMaxCliVers = Integer.MAX_VALUE;
+                        }
+
+                        // To older clients who can join, announce game without its options/version
+                        broadcastToVers(SOCNewGame.toCmd(gaName), gVers, newgameMaxCliVers);
+
+                        // To older clients who can't join, announce game with cant-join prefix
+                        StringBuffer sb = new StringBuffer();
+                        sb.append(SOCGames.MARKER_THIS_GAME_UNJOINABLE);
+                        sb.append(gaName);
+                        broadcastToVers
+                            (SOCNewGame.toCmd(sb.toString()), SOCGames.VERSION_FOR_UNJOINABLE, gVers-1);
+                    }
+                }
+            }
+            catch (IllegalArgumentException e)
+            {
+                if (! monitorReleased)
+                    gameList.releaseMonitor();
+                throw e;  // caller handles it
+            }
+            catch (Exception e)
+            {
+                D.ebugPrintStackTrace(e, "Exception in connectToGame");
+            }
+
+            if (!monitorReleased)
+            {
+                gameList.releaseMonitor();
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * the connection c leaves the game gm.  Clean up; if needed, call {@link #forceEndGameTurn(SOCGame, String)}.
+     *<P>
+     * WARNING: MUST HAVE THE gameList.takeMonitorForGame(gm) before
+     * calling this method
+     *
+     * @param c  the connection; if c is being dropped because of an error,
+     *           this method assumes that {@link StringConnection#disconnect()}
+     *           has already been called.  This method won't exclude c from
+     *           any communication about leaving the game, in case they are
+     *           still connected and in other games.
+     * @param gm the game
+     * @param gameListLock  true if we have the gameList.takeMonitor() lock
+     * @return true if the game was destroyed (because c was the last non-robot player,
+     *              and no one was watching)
+     */
+    public boolean leaveGame(StringConnection c, String gm, boolean gameListLock)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 51. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.